libvast/src/logical_pipeline.cpp

```diff
@@ -0,0 +1,420 @@
+  -> generator<runtime_batch> {
+  auto gen = op->make_runtime_physical_operator({}, ctrl);
+  if (not gen) {
+    ctrl->abort(gen.error());
+    return {};
+  }
+  auto f = []<element_type Input, element_type Output>(
+             physical_operator<Input, Output> gen) -> generator<runtime_batch> {
+    if constexpr (not std::is_void_v<Input>) {
+      die("tried to instantiate pipeline that does not begin with void");
+    } else {
+      // This coroutine captures keeps the physical operator alive.
+      for (auto&& output : gen()) {
+        VAST_INFO("yield output");
+        // This performs a conversion to `runtime_batch`.
+        co_yield std::move(output);
+      }
+    }
+  };
+  return std::visit(f, std::move(*gen));
+}
+
+struct instance_state {
+  generator<runtime_batch> gen;
+  generator<runtime_batch>::iterator it;
+  std::queue<runtime_batch> queue;
+  runtime_physical_operator physical_op;
+};
+
+template <class Batch>
+auto generator_for_queue(std::queue<runtime_batch>& queue, bool& stop)
+  -> generator<Batch> {
+  while (!stop) {
+    if (queue.empty()) {
+      // MESSAGE("yield empty");
+      co_yield Batch{};
+      continue;
+    }
+    auto batch_ptr = std::get_if<Batch>(&queue.front());
+    VAST_ASSERT(batch_ptr);
+    auto batch = std::move(*batch_ptr);
+    queue.pop();
+    VAST_ASSERT(batch_traits<Batch>::size(batch) != 0);
+    // MESSAGE("yield element");
+    co_yield std::move(batch);
+  }
+}
+
+template <class Batch>
+  requires(!std::same_as<Batch, runtime_batch>)
+auto erase_producer(generator<Batch> producer) -> generator<runtime_batch> {
+  for (auto&& output : producer) {
+    co_yield std::move(output);
+  }
+}
+
+template <class InputBatch>
+auto make_producer_from_queue(std::queue<runtime_batch>& queue, bool& stop,
+                              runtime_physical_operator& physical_op)
+  -> generator<runtime_batch> {
+  auto input_generator = generator_for_queue<InputBatch>(queue, stop);
+  auto make_generator
+    = [&]<element_type Input, element_type Output>(
+        physical_operator<Input, Output>& physical_op) noexcept
+    -> generator<runtime_batch> {
+    if constexpr (std::is_void_v<Input>) {
+      die("input type of operator must not be void here");
+    } else if constexpr (!std::is_same_v<element_type_to_batch_t<Input>,
+                                         InputBatch>) {
+      die("input generator did not yield operator input type");
+    } else {
+      return erase_producer(physical_op(std::move(input_generator)));
+    }
+  };
+  return std::visit(make_generator, physical_op);
+}
+
+auto exhaust_on_stall(instance_state& instance) -> generator<runtime_batch> {
+  // This effectively returns a generator that pulls from the instantiation,
+  // which pulls from the queue. A stall is requested when the queue is empty.
+  // In this case, the generator returned here becomes exhausted, which will
+  // request the next batch from `previous`.
```

Comment:
We should make it a post-condition that this always consumes the input batch. Then it will always be exactly one and we can use something more lightweight than a queue.

Reply:
While that may work in the local executor, we will need the queue once we move to an actor-based approach where a logical operator may be triggered multiple times in a row. Right now the queue size should never exceed 1, but I don't think it'll stay that way for long.

Reply:
I have cleaned this up, changed the behavior a bit, and switched to `std::optional` instead of `std::queue` for now. But we will have to consider this when implementing the actor-based executor.